src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,273 @@
+                switch (_reader.TokenType)
+                {
+                    case JsonTokenType.String:
+                        return new[] { (string)_reader.GetString() };
+
+                    case JsonTokenType.StartArray:
+                        return ReadStringArrayAsReadOnlyListFromArrayStart();
+
+                    case JsonTokenType.StartObject:
+                        return null;
+                }
+            }
+
+            return null;
+        }
+
+        internal IReadOnlyList<string> ReadStringArrayAsReadOnlyListFromArrayStart()
+        {
+            ThrowExceptionIfDisposed();
+
+            List<string> strings = null;
+
+            while (Read() && _reader.TokenType != JsonTokenType.EndArray)
+            {
+                string value = _reader.ReadTokenAsString();
+
+                strings = strings ?? new List<string>();
+
+                strings.Add(value);
+            }
+
+            return (IReadOnlyList<string>)strings ?? Array.Empty<string>();
+        }
+
+        // This function is called when Read() returns false and we're not already in the final block
+        private void GetMoreBytesFromStream()
+        {
+            if (_reader.BytesConsumed < _bufferUsed)
+            {
+                // If the number of bytes consumed by the reader is less than the amount set in the buffer then we have leftover bytes
+                var oldBuffer = _buffer;
+                ReadOnlySpan<byte> leftover = oldBuffer.AsSpan((int)_reader.BytesConsumed);
+                _bufferUsed = leftover.Length;
+
+                // If the leftover bytes are the same as the buffer size then we are at capacity and need to double the buffer size
+                if (leftover.Length == _buffer.Length)
+                {
+                    _buffer = _bufferPool.Rent(_buffer.Length * 2);
+                    leftover.CopyTo(_buffer);
+                    _bufferPool.Return(oldBuffer, true);
+                }
+                else
+                {
+                    leftover.CopyTo(_buffer);
+                }
+            }
+            else
+            {
+                _bufferUsed = 0;
+            }
+
+            ReadStreamIntoBuffer(_reader.CurrentState);
+        }
+
+        /// <summary>
+        /// Loops through the stream and reads it into the buffer until the buffer is full or the stream is empty, creates the Utf8JsonReader. 
+        /// </summary>
+        private void ReadStreamIntoBuffer(JsonReaderState jsonReaderState)
+        {
+            int bytesRead;
+            do
+            {
+                var spaceLeftInBuffer = _buffer.Length - _bufferUsed;
+                bytesRead = _stream.Read(_buffer, _bufferUsed, spaceLeftInBuffer);
+                _bufferUsed += bytesRead;
+            }
+            while (bytesRead != 0 && _bufferUsed != _buffer.Length);
+            _reader = new Utf8JsonReader(_buffer.AsSpan(0, _bufferUsed), isFinalBlock: bytesRead == 0, jsonReaderState);
```

Comment:
https://learn.microsoft.com/dotnet/standard/serialization/system-text-json/use-utf8jsonreader doc suggests to use `_reader.CurrentState`. If this suggestion is valid for our scenario, then we can remove the `jsonReaderState` parameter from the method definition.

```suggestion
            _reader = new Utf8JsonReader(_buffer.AsSpan(0, _bufferUsed), isFinalBlock: bytesRead == 0, _reader.CurrentState);
```

Reply:
The function is taking in the jsonReaderState because in the constructor the state doesn't already exist.